drivers/net/ethernet/nuvoton/npcm7xx_emc.c

```diff
@@ -925,19 +926,41 @@ static void npcm7xx_return_default_idle(struct net_device *netdev)
 	 * register that its reset value is not 0,
 	 * we choose (ether->reg + REG_FFTCR).
 	 */
+	deadline = ktime_add_ms(ktime_get(), 20);
 	do {
 		val = readl((ether->reg + REG_FFTCR));
-	} while (val == 0);
+		if (val != 0)
+			break;
+		cpu_relax();
+	} while (ktime_before(ktime_get(), deadline));
+
+	if (val == 0) {
+		dev_err(&ether->pdev->dev,
+			"Timed out waiting for FFTCR to become nonzero.\n");
+		return -ETIMEDOUT;
+	}
 
 	/* Now we can verify if (ether->reg + REG_MCMDR).SWR became
 	 * 0 (probably it will be 0 on the first read).
 	 */
+	deadline = ktime_add_ms(ktime_get(), 20);
 	do {
 		val = readl((ether->reg + REG_MCMDR));
-	} while (val & SWR);
+		if (!(val & SWR))
+			break;
+		cpu_relax();
```

Comment:
I don't like that on first reading of 0 it will do cpu_relax() since I know that the time it will read 0 is very short, sould be few microseconds or even less than 1us.
I prefer a busy loop of 10us and every 10us go to cpu_relax().